druid/src/widget/tabs.rs

```diff
@@ -0,0 +1,960 @@
+// Copyright 2020 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget that can switch between one of many views, hiding the inactive ones.
+//!
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::rc::Rc;
+
+use crate::kurbo::Line;
+use crate::piet::RenderContext;
+
+use crate::widget::{Axis, CrossAxisAlignment, Flex, Label, LensScopeTransfer, Scope, ScopePolicy};
+use crate::{
+    theme, Affine, BoxConstraints, Data, Env, Event, EventCtx, Insets, LayoutCtx, Lens, LifeCycle,
+    LifeCycleCtx, PaintCtx, Point, Rect, SingleUse, Size, UpdateCtx, Widget, WidgetExt, WidgetPod,
+};
+use instant::Duration;
+
+type TabsScope<TP> = Scope<TabsScopePolicy<TP>, Box<dyn Widget<TabsState<TP>>>>;
+type TabBodyPod<TP> = WidgetPod<<TP as TabsPolicy>::Input, <TP as TabsPolicy>::BodyWidget>;
+type TabBarPod<TP> = WidgetPod<TabsState<TP>, Box<dyn Widget<TabsState<TP>>>>;
+type TabIndex = usize;
+type Nanos = u64; // TODO: Make Duration Data?
+
+pub struct TabInfo {
+    pub name: String,
+    pub can_close: bool,
+}
+
+impl TabInfo {
+    pub fn new(name: String, can_close: bool) -> Self {
+        TabInfo { name, can_close }
+    }
+}
+
+/// A policy that determines how a Tabs instance derives its tabs from its app data
+pub trait TabsPolicy: Data {
+    /// The identity of a tab.
+    type Key: Hash + Eq + Clone + Debug;
+
+    /// The input data that will a) be used to derive the tab and b) also be the input data of all the child widgets.
+    type Input: Data;
+
+    /// The common type for all body widgets in this set of tabs.
+    /// A flexible default is Box<dyn Widget<Self::Input>>
+    type BodyWidget: Widget<Self::Input>;
+
+    /// The common type for all label widgets in this set of tabs
+    /// Usually this would be Label<Self::Input>
+    type LabelWidget: Widget<Self::Input>;
+
+    /// This policy whilst it is being built.
+    /// This is only be useful for implementations supporting AddTab, such as StaticTabs.
+    /// It can be filled in with () by other implementations until associated type defaults are stable
+    type Build;
+
+    /// Have the tabs changed. Expected to be cheap, eg pointer or numeric comparison.
+    fn tabs_changed(&self, old_data: &Self::Input, data: &Self::Input) -> bool;
+
+    /// What are the current tabs set in order.
```

Comment:
?

Reply:
Left over comment from a method I got rid of